2020/day21/solution.py

```diff
@@ -0,0 +1,43 @@
+from __future__ import annotations
+from collections import Counter
+import re
+
+food_re = re.compile(r"(?P<ingredients>.*) \(contains (?P<allergens>.*)\)")
+
+def get_ingredients_and_allergens(line: str) -> tuple[set[str], set[str]]:
+    match = food_re.match(line)
+    ingredients = set(ingredient.strip() for ingredient in match['ingredients'].split(' '))
+    allergens = set(ingredient.strip() for ingredient in match['allergens'].split(','))
+    return (ingredients, allergens)
+
+def produce_allergen_list(allergen_map: dict[str, set[str]]) -> str:
```

Comment:
Couple comments:
- best not to mutate the input dictionary; I did some local edits that a) use a dict comprehension replacing the second loop and parts of the first to do any needed mutations b) replaced `.pop()` with a faster nonmutating alternative.
- return a set of ingredients directly; the caller should be responsible for formatting (e.g. sorting, string combining)
- better names would be helpful in reading (had to rewrite locally with more precise names to understand)